pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -106,6 +120,53 @@ func haveOverlap(a1, a2 []string) bool {
 	return false
 }
 
+func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	return pl.isReadWriteOncePodAccessModeConflict(pod)
+}
+
+// isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
+// This access mode restricts volume access to a single pod on a single node. Since only a single pod can
+// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as unschedulable and unresolvable.
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(pod *v1.Pod) *framework.Status {
+	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
+	if err != nil {
+		return framework.NewStatus(framework.Error, ErrReasonNodeInfoError)
+	}
+
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(volume.PersistentVolumeClaim.ClaimName)
+		if err != nil {
+			return framework.NewStatus(framework.Error, ErrReasonPVCError)
+		}
+
+		// If ReadWriteOncePod is used, it can be the only one on the PVC since it overrides any
+		// additional access modes. This is also because we do not know what access mode the pod
+		// intends to use if there are multiple.
+		if !(v1helper.ContainsAccessMode(pvc.Spec.AccessModes, v1.ReadWriteOncePod) && len(pvc.Spec.AccessModes) == 1) {
+			continue
+		}
+
+		key := pod.Namespace + "/" + volume.PersistentVolumeClaim.ClaimName
+		for _, nodeInfo := range nodeInfos {
+			refCount, ok := nodeInfo.PVCRefCounts[key]
+			// Enforce ReadWriteOncePod access mode. This is also enforced during volume mount in kubelet.
+			if ok && utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod) && refCount > 0 {
```

Comment:
I'd say, let's not check the feature gate in scheduler.
The apiserver would already remove/disallow the feature at pod creation time. That is generally enough for pods.

Reply:
Also, you can get rid of `ok`. `refCount` will be 0 by default.

Reply:
+1 to removing the feature gate check, or use the pattern described here: https://github.com/kubernetes/kubernetes/blob/a3f24e8459465495738af1b9cc6c3db80696e3c1/pkg/scheduler/framework/plugins/registry.go#L52 and do the check in outside this whole function (skip the whole function if false, why wait until we arrive here?)

Reply:
Done. I opted to use the feature gate, just to have the flexibility to experiment with enforcing the access mode in both the scheduler and/or kubelet